core/genesis.go

```diff
@@ -596,28 +596,44 @@ func DefaultHoleskyGenesisBlock() *Genesis {
 }
 
 // DeveloperGenesisBlock returns the 'geth --dev' genesis block.
-func DeveloperGenesisBlock(gasLimit uint64, faucet common.Address) *Genesis {
+func DeveloperGenesisBlock(gasLimit uint64, faucet common.Address, genesisAlloc *GenesisAlloc) *Genesis {
 	// Override the default period to the user requested one
 	config := *params.AllDevChainProtocolChanges
+	alloc := map[common.Address]GenesisAccount{
+		common.BytesToAddress([]byte{1}): {Balance: big.NewInt(1)}, // ECRecover
+		common.BytesToAddress([]byte{2}): {Balance: big.NewInt(1)}, // SHA256
+		common.BytesToAddress([]byte{3}): {Balance: big.NewInt(1)}, // RIPEMD
+		common.BytesToAddress([]byte{4}): {Balance: big.NewInt(1)}, // Identity
+		common.BytesToAddress([]byte{5}): {Balance: big.NewInt(1)}, // ModExp
+		common.BytesToAddress([]byte{6}): {Balance: big.NewInt(1)}, // ECAdd
+		common.BytesToAddress([]byte{7}): {Balance: big.NewInt(1)}, // ECScalarMul
+		common.BytesToAddress([]byte{8}): {Balance: big.NewInt(1)}, // ECPairing
+		common.BytesToAddress([]byte{9}): {Balance: big.NewInt(1)}, // BLAKE2b
+	}
+
+	if genesisAlloc != nil {
+		precompiled := map[common.Address]struct{}{}
+		for addr := range alloc {
+			precompiled[addr] = struct{}{}
+		}
+		for addr, account := range *genesisAlloc {
+			// Don't overwrite the precompiled accounts
+			if _, ok := precompiled[addr]; ok {
+				log.Crit("DeveloperGenesisBlock: can't overwrite precompiled account", "address", addr)
+			}
+			alloc[addr] = account
+		}
```

Comment:
How about a bit simpler:
```suggestion
		for addr, account := range *genesisAlloc {
			// Don't allow overwriting the predefined accounts (precompiles)
			if _, ok := alloc[addr]; ok {
				log.Crit("Overwriting precompiles disallowed", "address", addr)
			}
			alloc[addr] = account
		}
```

Reply:
In this case, an error will also be reported if there is a duplicate address in the genesis file specified by the user, eg:

```json
{
  "0x00dc25665018aa1fe0e6bc666dac8fc2697ff9ba": {
    "balance": "0x150ca"
  },

  "0x00dc25665018aa1fe0e6bc666dac8fc2697ff9ba": {
    "balance": "0x150cb"
  }
}

```

Reply:
Seems the duplicated case will not happen, Golang JSON unmarshal will ignore the first one.

Reply:
Right, or rather, I don't think it's about json: it's now iteration over maps work. Internally, it picks out the set of keys to iterate over, and that set does not mutate during the course of the iteration. I remember having seen this before. 

I am 95% certain that it is safe to add/remove keys from maps that are being iterated on, without it affecting the iteration.